### PR TITLE
 Give broadcast_coalesced tensors different version counters

### PR DESCRIPTION
**Pull request by ssnl**
In `broadcast_coalesced`, since multiple variables can be "views" of a big flattened tensor, they can share the same version counter. However, this base flat tensor is not exposed and they don't share any memory locations, so this is not necessary. Furthermore, it can cause problems, e.g., when two buffers are broadcast together in `DataParallel` and one of them is modified in-place during `forward` but the other is needed in backward, autograd engine will complain.

Fixing the bug discovered at https://github.com/pytorch/pytorch/pull/13350#issuecomment-436011370

edit: This is a very real problem. E.g., consider using Spectral Norm + Batch Norm together.

### CODE REVIEW

File: torch/csrc/cuda/comm.cpp [hunk 1]
```diff
@@ -97,17 +129,25 @@ tensor_list2d broadcast_coalesced(TensorList tensors, IntList devices, size_t bu
         auto & device_outputs = outputs[i];
         auto & inds = broadcast_indices[i];
         auto & vals = broadcast_values[i];
-        for (auto & t : utils::unflatten_sparse_tensors(inds, vals, chunk.tensors))
-          device_outputs.push_back(std::move(t));
+        for (auto & t : utils::unflatten_sparse_tensors(inds, vals, chunk.tensors)) {
+          // See NOTE [ Version Counter in comm.*_coalesced ]
+          AT_ASSERT(t.is_variable());
+          Variable var = t;
+          device_outputs.push_back(std::move(make_variable(var.data(), false)));
```
Comment (colesbury) [inline torch/csrc/cuda/comm.cpp]: Remove the `std::move()`. `make_variable` already returns an r-value.

File: torch/csrc/cuda/comm.cpp [hunk 2]
```diff
@@ -97,17 +129,25 @@ tensor_list2d broadcast_coalesced(TensorList tensors, IntList devices, size_t bu
...
       for (size_t i = 1, num_devices = devices.size(); i < num_devices; ++i) {
         device_guard.set_index(devices[i]);
         auto & device_outputs = outputs[i];
-        for (auto & t : utils::unflatten_dense_tensors(results[i], chunk.tensors))
-          device_outputs.push_back(std::move(t));
+        for (auto & t : utils::unflatten_dense_tensors(results[i], chunk.tensors)) {
+          // See NOTE [ Version Counter in comm.*_coalesced ]
+          AT_ASSERT(t.is_variable());
+          Variable var = t;
+          device_outputs.push_back(std::move(make_variable(var.data(), false)));
```
Comment (colesbury) [inline torch/csrc/cuda/comm.cpp]: remove std::move

**Review by colesbury — APPROVED** [review]
